### PR TITLE
Fortran backend precision support (#578)

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for review from either @arporter or @sergisiso.  Note, there is one question in the text from which I would appreciate feedback on what to do.

File: src/psyclone/psyir/backend/fortran.py [hunk 1]
```diff
@@ -109,6 +109,79 @@ def gen_dims(symbol):
...
+        datatype = "logical"
+    else:
+        datatype = symbol.datatype
+    if not symbol.precision:
+        # This symbol has no precision information so simply return
+        # the name of the datatype.
+        return datatype
+    if isinstance(symbol.precision, int):
+        if datatype not in ['real', 'integer', 'logical']:
+            raise VisitorError("Explicit precision not support for datatype "
```
Comment (arporter) [inline src/psyclone/psyir/backend/fortran.py]: "supported"

File: src/psyclone/psyir/backend/fortran.py [hunk 2]
```diff
@@ -109,6 +109,79 @@ def gen_dims(symbol):
...
+
+    if isinstance(symbol.precision, Symbol.Precision):
+        # The precision information is not absolute so is either
+        # machine specific or is specified via the compiler. Fortran
+        # only distinguishes relative precision for single and double
+        # precision reals.
+        if datatype.lower() == "real" and \
+           symbol.precision == Symbol.Precision.DOUBLE:
+            return "double precision"
+        import logging
```
Comment (arporter) [inline src/psyclone/psyir/backend/fortran.py]: As discussed, although this is a clear use case for logging, we shouldn't just introduce it here without also adding top-level controls to the driver script. Therefore please comment it out for now and add a reference to the logging Issue (#11 ).

File: src/psyclone/psyir/backend/fortran.py [hunk 3]
```diff
@@ -109,6 +109,79 @@ def gen_dims(symbol):
...
+        logging.warning(
+            ("Fortran does not support relative precision for the '%s' "
+             "datatype but '%s' was specified for variable '%s'.",
+             datatype, str(symbol.precision), symbol.name))
+        return datatype
+    if isinstance(symbol.precision, Symbol):
+        if datatype not in ["real", "integer", "logical"]:
+            raise VisitorError("kind not support for datatype "
```
Comment (arporter) [inline src/psyclone/psyir/backend/fortran.py]: "supported"

File: src/psyclone/psyir/backend/fortran.py [hunk 4]
```diff
@@ -109,6 +109,79 @@ def gen_dims(symbol):
...
+                               "'{0}' in Fortran backend.".format(datatype))
+        # Question: Should we check whether the value of precision is
+        # known and, if so, what should we do if not?
+        # The precision information is provided by a parameter, so use KIND.
```
Comment (arporter) [inline src/psyclone/psyir/backend/fortran.py]: As discussed, I think we should opt for lazy evaluation here and not attempt to discover things that we don't actually need to know.

File: src/psyclone/psyir/backend/fortran.py (hunk 2 above)
Comment (arporter) [reply src/psyclone/psyir/backend/fortran.py]: (Also, you have superfluous parentheses in the `logging.warning` call.

File: src/psyclone/psyir/backend/fortran.py [hunk 5]
```diff
@@ -109,6 +109,79 @@ def gen_dims(symbol):
     return dims
 
 
+def gen_datatype(symbol):
+    '''Given a Symbol instance as input, return the datatype of the
```
Comment (arporter) [inline src/psyclone/psyir/backend/fortran.py]: Please include the name of the Symbol in the error messages produced by this routine.

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 6]
```diff
@@ -111,6 +112,115 @@ def test_gen_dims_error(monkeypatch):
     assert "unsupported gen_dims index 'invalid'" in str(excinfo)
 
 
+def test_gen_datatype():
+    '''Check the gen_datatype function produces the expected datatypes.'''
+    for datatype in ["real", "integer", "character"]:
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: `@pytest.mark.parametrize`?

Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Actually, you could fold this test into the following one?

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 7]
```diff
@@ -111,6 +112,115 @@ def test_gen_dims_error(monkeypatch):
...
+    precision is specified.
+
+    '''
+    symbol = Symbol("dummy", datatype, precision=precision)
+    assert gen_datatype(symbol) == result
+
+
+# Note, in isolation and when only running tests from the psyir
+# directory, this test does not fail.
+@pytest.mark.xfail(reason="issue #582 backend logging output is affected by "
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: I don't think there's a need for a new issue - just refer to #11 here as it's something that will need to be addressed when we implement logging.

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 8]
```diff
@@ -111,6 +112,115 @@ def test_gen_dims_error(monkeypatch):
...
+    # create logicals with a precision
+    monkeypatch.setattr(symbol, "precision", 32)
+    with pytest.raises(VisitorError) as excinfo:
+        _ = gen_datatype(symbol)
+    assert ("Datatype 'logical' supports fixed precision of [1, 2, 4, 8, 16] "
+            "but found '32'." in str(excinfo.value))
+
+    # Kind not supported for character
+    symbol = Symbol("dummy", "real", precision=Symbol("c_def", "integer"))
+    # This needs to be monkeypatched as the Symbols can not
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: "the Symbol constructor cannot"?

File: src/psyclone/psyir/backend/fortran.py (hunk 5 above)
Comment (arporter) [reply src/psyclone/psyir/backend/fortran.py]: Also, this routine feels a bit long. Is it worth creating some internal routines that handle e.g. precision=int, precision=Symbol separately?

File: src/psyclone/psyir/backend/fortran.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/fortran.py]: Oops.

File: src/psyclone/psyir/backend/fortran.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/fortran.py]: Oh, alright then. As agreed I'll keep the code for future reference, but comment it out.

File: src/psyclone/psyir/backend/fortran.py (hunk 3 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/fortran.py]: Oops again, probably cut-n-paste.

File: src/psyclone/psyir/backend/fortran.py (hunk 5 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/fortran.py]: Symbol names added and I also forgot to document the exceptions so have done this too. I've not yet looked at splitting the function.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 7 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: OK. I've changed the reference and commented out the test.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 8 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Updated.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 6 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: I've used parametrize. The test can't be folded into the next one as this test is without the optional precision argument and the next one is with it.

File: src/psyclone/psyir/backend/fortran.py (hunk 5 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/fortran.py]: I've looked at splitting the routine and don't think it would help to be honest.

**Comment by rupertford** [comment]
The broken compile example now works which is good news. I needed to change `use constants_mod` to `use constants_mod, only : r_def` in the two relevant kernel files. The reason for this is that, at the moment, use statements without an only clause do not appear in the output. An additional issue is that an `integer : r_def` variable is declared unless the `use constants_mod, only : r_def` is specified in the subroutine. For some reason the compiler does not seem to mind this, although I'm not sure we would get the expected behaviour. In lfric kind constants are declared in the module so we will need to check ancestor symbol tables, not just the parent.

**Comment by rupertford** [comment]
Ready for next review from @arporter subject to travis being happy.

**Pull request merged**